drivers: dma support for the stm32u5x

FRASTM opened the pull request:
This PR is introducing the support of the DMA  from the GPDMA peripheral on the stm32u5 mcu

The mem-to-mem test case are provided 
- tests/drivers/dma/loop_transfer
- tests/drivers/dma/chan_blen_transfer
 to be executed on the b_u585i_iot02a disco kit  or nucleo_u575zi_q board

Signed-off-by: Francois Ramu [francois.ramu@st.com](mailto:francois.ramu@st.com)

FRASTM committed:
dts: bindings: dma: introduce a new type of DMA

This commit defines a new dma version for devices like stm32U5.
The peripheral is a GPDMA in this soc serie.
It has several specific definitions used by its stm32 LL driver
compared to the V2, including up to 16 channels.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: dma: new dma peripheral driver for stm32u5

It includes the definition for the DMA peripheral of this type,
present on stm32U5 devices.
A particular DMA_STM32U5 config is selected for that purpose.
The driver is derived from the existing dma_stm32.c
The GPDMA is counting channels (0-15) ; stream offset is 0.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32U5 add a DMA node for the GPDMA1 instance

This commit defines the GPDMA peripheral for the stm32U5.
This dma is of a new type with 16 channels and 114 requests.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 disco kit target board has GPDMA instance

Enable the GPDMA 1 instance on the target board b_u585i_iot02a
from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
```
*** Booting Zephyr OS build zephyr-v3.0.0-3298-g0fafe0910b8d  ***
Running test suite dma_m2m_loop_test
===================================================================
START - test_dma_m2m_loop
DMA memory to memory transfer started on DMA_1                                                       
Preparing DMA Controller                                                                             
this platform do not support the dma channel                                                         
Starting the transfer on channel 2 and waiting for 1 second                                          
Each RX buffer should contain the full TX buffer string.                                             
RX data Loop 0: The quick brown fox jumps over the lazy dog                                          
...                                               
RX data Loop 1: The quick brown fox jumps over the lazy dog                                          
...                                                    
RX data Loop 2: The quick brown fox jumps over the lazy dog                                          
...                                                     
Finished: DMA                                                                                        
 PASS - test_dma_m2m_loop in 0.606 seconds                                                           
===================================================================                                  
START - test_dma_m2m_loop_suspend_resume                                                             
DMA memory to memory transfer started on DMA_1                                                       
Preparing DMA Controller                                                                             
this platform do not support the dma channel                                                         
Starting the transfer on channel 2 and waiting for 1 second                                          
suspend not supported                                                                                
 PASS - test_dma_m2m_loop_suspend_resume in 0.18 seconds                                             
===================================================================                                  
Test suite dma_m2m_loop_test succeeded                                                               
===================================================================                                  
PROJECT EXECUTION SUCCESSFUL 
```

FRASTM commented:
```
*** Booting Zephyr OS build zephyr-v3.0.0-3298-g0fafe0910b8d  ***                                    
Running test suite dma_m2m_test                                                                      
===================================================================                                  
START - test_dma_m2m_chan0_burst8                                                                    
Preparing DMA Controller: Chan_ID=5, BURST_LEN=1                                                     
Starting the transfer                                                                                
DMA transfer done                                                                                    
It is harder to be kind than to be wise........                                                      
 PASS - test_dma_m2m_chan0_burst8 in 2.13 seconds                                                    
===================================================================                                  
START - test_dma_m2m_chan1_burst8                                                                    
Preparing DMA Controller: Chan_ID=9, BURST_LEN=1                                                     
Starting the transfer                                                                                
DMA transfer done                                                                                    
It is harder to be kind than to be wise........                                                      
 PASS - test_dma_m2m_chan1_burst8 in 2.13 seconds                                                    
===================================================================                                  
START - test_dma_m2m_chan0_burst16                                                                   
Preparing DMA Controller: Chan_ID=5, BURST_LEN=2                                                     
Starting the transfer                                                                                
DMA transfer done                                                                                    
It is harder to be kind than to be wise........                                                      
 PASS - test_dma_m2m_chan0_burst16 in 2.13 seconds                                                   
===================================================================                                  
START - test_dma_m2m_chan1_burst16                                                                   
Preparing DMA Controller: Chan_ID=9, BURST_LEN=2                                                     
Starting the transfer                                                                                
DMA transfer done                                                                                    
It is harder to be kind than to be wise........                                                      
 PASS - test_dma_m2m_chan1_burst16 in 2.13 seconds                                                   
===================================================================                                  
Test suite dma_m2m_test succeeded                                                                    
===================================================================                                  
PROJECT EXECUTION SUCCESSFUL        
```

erwango commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,84 @@
+  It is present on stm32 devices stm32U5
+  This DMA controller includes several channels with different requests.
+  DMA clients connected to the STM32 DMA controller must use the format
+  described in the dma.txt file, using a four-cell specifier for each
```
dma.txt file doesn't exist.

FRASTM replied:
line removed

FRASTM replied:
Same desrciption exists for all the st,stm32-dma-v_X_.yaml

erwango commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,84 @@
+  DMA clients connected to the STM32 DMA controller must use the format
+  described in the dma.txt file, using a four-cell specifier for each
+  capable of supporting 15 dependent DMA channels.
+  DMA clients connected to the STM32 DMA controller must use the format
+  described in the dma.txt file, using a 3-cell specifier for each
```
?

erwango commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,84 @@
+               0x1: medium
+               0x2: high
+               0x3: very high
+  Example of dma node for stm32wb55x
+     dma2: dma-controller@40020400 {
+         compatible = "st,stm32-dma-v2";
+         ...
+         dma-requests = <7>;
+         status = "disabled";
+         label = "DMA_2";
+        };
```
Let's remove this. It doesn't help much users to use it.

FRASTM replied:
example removed

erwango commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,84 @@
+         label = "DMA_2";
+        };
+
+  For the client part, example for stm32l476 on DMA1 instance
+    Tx using channel 3 with request 1
+    Rx using channel 2 with request 1
+    spi1 {
+     compatible = "st,stm32-spi";
+     dmas = <&dma1 3 1 0x20440>,
+            <&dma1 2 1 0x20480>;
+     dma-names = "tx", "rx";
+    };
```
Please update example to fit this v3 binding

FRASTM replied:
done

erwango commented on dts/bindings/dma/st,stm32-dma-v3.yaml:
```diff
@@ -0,0 +1,84 @@
+
+# Parameter syntax of stm32 follows the dma client dts syntax
+# in the Linux kernel declared in
+# https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/st,stm32-dma.yaml
```
It doesn't look to be the case anymore, so let's remove this.
Also, comments are not rendered in on line doc.

FRASTM replied:
removed

erwango commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -399,6 +399,20 @@
+			dma-requests = <114>;
+			dma-offset = <0>;
+			status = "disabled";
+			label = "DMA_1";
```
Should it be GPDMA1 ?

FRASTM replied:
DMA is well known, GPDMA could be too specific,  and moreover it's _just_ a label

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -93,6 +109,11 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
 
 	stream = &config->streams[id];
 
+	/* Do not continue if dma channel is not busy anymore */
+	if (stream->busy == false) {
+		return;
+	}
+
```
If affecting existing DMAs, please move it to a dedicated PR.
(In case of revert is required for instance, but also to allow users that are interested in non v3 dmas to review)

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -358,6 +412,12 @@ DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
 	}
 
+#if defined(CONFIG_DMA_STM32_V3)
+	DMA_InitStruct.SrcAddress =	config->head_block->source_address;
```
```suggestion
	DMA_InitStruct.SrcAddress = config->head_block->source_address;
```

FRASTM replied:
done

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -632,7 +774,11 @@ DMA_STM32_EXPORT_API int dma_stm32_get_status(const struct device *dev,
 
 	stream = &config->streams[id];
+#if defined(CONFIG_DMA_STM32_V3)
+	stat->pending_length = LL_DMA_GetBlkDataLength(dma, dma_stm32_id_to_stream(id));
+#else
 	stat->pending_length = LL_DMA_GetDataLength(dma, dma_stm32_id_to_stream(id));
```
ditto

FRASTM replied:
changed with a redefiniion of the LL_DMA_GetBlkDataLength for the stm32U5x, in the header file

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -516,23 +641,40 @@ DMA_STM32_EXPORT_API int dma_stm32_reload(const struct device *dev, uint32_t id,
 	if (stream->source_periph) {
+#if defined(CONFIG_DMA_STM32_V3)
+		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id),
+#else
 		LL_DMA_SetDataLength(dma, dma_stm32_id_to_stream(id),
-				     size / stream->src_size);
+#endif /* CONFIG_DMA_STM32_V3 */
```
To maintain code readability, please add new `dma_stm32_foo` functions to abstract these LL API differences.

FRASTM replied:
changed  with  a redefiniion of the LL_DMA_SetBlkDataLength for the stm32U5x, in the header file

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -516,23 +641,40 @@ DMA_STM32_EXPORT_API int dma_stm32_reload(const struct device *dev, uint32_t id,
 		break;
 	case MEMORY_TO_MEMORY:
 	case PERIPHERAL_TO_MEMORY:
+#if defined(CONFIG_DMA_STM32_V3)
+		LL_DMA_ConfigAddresses(dma, dma_stm32_id_to_stream(id), dst, src);
+#else
 		LL_DMA_SetPeriphAddress(dma, dma_stm32_id_to_stream(id), src);
 		LL_DMA_SetMemoryAddress(dma, dma_stm32_id_to_stream(id), dst);
+#endif /* CONFIG_DMA_STM32_V3 */
```
To maintain code readability, please add new `dma_stm32_foo` functions to abstract these LL API differences.

FRASTM replied:
adding `stm32_dma_set_mem_periph_address()` and `stm32_dma_set_periph_mem_address()`
to abstract the LL driver

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -516,23 +641,40 @@ DMA_STM32_EXPORT_API int dma_stm32_reload(const struct device *dev, uint32_t id,
+#endif /* CONFIG_DMA_STM32_V3 */
+			    size / stream->src_size);
 	} else {
+#if defined(CONFIG_DMA_STM32_V3)
+		LL_DMA_SetBlkDataLength(dma, dma_stm32_id_to_stream(id),
+#else
 		LL_DMA_SetDataLength(dma, dma_stm32_id_to_stream(id),
+#endif /* CONFIG_DMA_STM32_V3 */
```
ditto

FRASTM replied:
macro do redefine LL_DMA_SetDataLength  and LL_DMA_GetDataLength in case of stm32u5xx LL

erwango reviewed (changes requested):
Looks good in general.
My main request would be to work a bit on abstraction in order to maintain driver code readability and avoid adding tooo much #ifdefs. I've provided suggestion in that sense.
Maybe it would also be required to provide abstraction on `dma_stm32_configure` which is now cluttered by  #ifdef's

FRASTM commented:
adding stm32_dma_set_mem_periph_address() and stm32_dma_set_periph_mem_address()
to abstract the LL driver in case of stm32U5 
--> stm32u5xx_ll_dma.h differs in several function names from other series.

FRASTM commented:
running tests/drivers/dma/ : 
```
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_f746zg,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_wb55rg,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,stm32l562e_dk,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_g474re,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_wl55jc,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,stm32f3_disco,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_f429zi,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_l152re,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,b_u585i_iot02a,passed
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_f091rc,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,nucleo_f103rb,passed,
tests/drivers/dma/loop_transfer/drivers.dma.loop_transfer,arm,disco_l475_iot1,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_f746zg,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_wb55rg,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,stm32l562e_dk,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_g474re,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_wl55jc,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,stm32f3_disco,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_f429zi,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_l152re,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,b_u585i_iot02a,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_f091rc,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,nucleo_f103rb,passed,
tests/drivers/dma/chan_blen_transfer/drivers.dma.chan_blen_transfer,arm,disco_l475_iot1,passed,
```

FRASTM commented:
rebase on 90537fe447d66e627b62efaa26f28d7a18926adc

FRASTM commented:
Correction in the source/dest adresses during the dma_reload : put in the same order as in the dma_config

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -158,16 +174,32 @@ static int dma_stm32_get_priority(uint8_t priority, uint32_t *ll_priority)
 {
 	switch (priority) {
 	case 0x0:
+#if defined(CONFIG_DMA_STM32_V3)
+		*ll_priority = LL_DMA_LOW_PRIORITY_LOW_WEIGHT;
+#else
 		*ll_priority = LL_DMA_PRIORITY_LOW;
+#endif /* CONFIG_DMA_STM32_V3 */
 		break;
 	case 0x1:
+#if defined(CONFIG_DMA_STM32_V3)
+		*ll_priority = LL_DMA_LOW_PRIORITY_MID_WEIGHT;
+#else
 		*ll_priority = LL_DMA_PRIORITY_MEDIUM;
+#endif /* CONFIG_DMA_STM32_V3 */
 		break;
 	case 0x2:
+#if defined(CONFIG_DMA_STM32_V3)
+		*ll_priority = LL_DMA_LOW_PRIORITY_HIGH_WEIGHT;
+#else
 		*ll_priority = LL_DMA_PRIORITY_HIGH;
+#endif /* CONFIG_DMA_STM32_V3 */
 		break;
 	case 0x3:
+#if defined(CONFIG_DMA_STM32_V3)
+		*ll_priority = LL_DMA_HIGH_PRIORITY;
+#else
 		*ll_priority = LL_DMA_PRIORITY_VERYHIGH;
+#endif /* CONFIG_DMA_STM32_V3 */
```
Can we use an array or something else to get rid of all these #ifdef's ?

FRASTM replied:
that could be a new table
```
static uint32_t table_priority[4] = {
#if defined(CONFIG_DMA_STM32_V3)
	LL_DMA_LOW_PRIORITY_LOW_WEIGHT,
	LL_DMA_LOW_PRIORITY_MID_WEIGHT,
	LL_DMA_LOW_PRIORITY_HIGH_WEIGHT,
	LL_DMA_HIGH_PRIORITY,
#else
	LL_DMA_PRIORITY_LOW,
	LL_DMA_PRIORITY_MEDIUM,
	LL_DMA_PRIORITY_HIGH,
	LL_DMA_PRIORITY_VERYHIGH,
#endif /* CONFIG_DMA_STM32_V3 */
};
```
used by the dma_stm32_get_priority() function as follows:

```
	if (priority > ARRAY_SIZE(table_priority)) {
		LOG_ERR("Priority error. %d", priority);
		return -EINVAL;
	}

	*ll_priority = table_priority[priority];
	return 0;
```

FRASTM commented:
rebase
Change to have a **dma_stm32u5.c** dma driver  dedicated to the stm32U5 family
The bindings is now `st_stm32u5_dma`  with a particular DMA_STM32U5 config which differs from the existing DMA_STM32 config

erwango commented on dts/bindings/dma/st,stm32u5-dma.yaml:
```diff
@@ -0,0 +1,65 @@
+               0x2: high
+               0x3: very high
+
+   For the client part, example for stm32u585 on GPDMA1 instance
+    Tx using channel 0 with request 7
+    Rx using channel 1 with request 6
+    spi1 {
+     dmas = <&gpdma1 0 7 0x10440>,
+            <&gpdma1 1 6 0x10480>;
+     dma-names = "tx", "rx";
+    };
```
Let's move this under "  DMA clients connected to the STM32 DMA controller must use a three-cell
  specifier for each channel"

FRASTM replied:
moved

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -16,6 +18,13 @@ config DMA_STM32
+config DMA_STM32U5
+	bool "STM32 DMA driver for stm32U5"
+	select USE_STM32_LL_DMA
+	depends on SOC_FAMILY_STM32
```
Let's be more specific and use DT_COMPAT_ST_STM32U5_DMA

FRASTM replied:
Ok

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -16,6 +18,13 @@ config DMA_STM32
 	help
 	  DMA driver for STM32 series SoCs.
 
+config DMA_STM32U5
+	bool "STM32 DMA driver for stm32U5"
+	select USE_STM32_LL_DMA
+	depends on SOC_FAMILY_STM32
+	help
+	  Enable DMA support mainly for stm32U5 family.
+
```
Should it be moved under `if DMA_STM32`

FRASTM replied:
changed

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dma_stm32, CONFIG_DMA_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_dma)
+#define DT_DRV_COMPAT st_stm32u5_dma
+#endif
```
```suggestion
#define DT_DRV_COMPAT st_stm32u5_dma
```

FRASTM replied:
removing #if DT_HAS_COMPAT_STATUS_OKAY(st_stm32u5_dma)

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+#define DT_DRV_COMPAT st_stm32u5_dma
+#endif
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+#if DT_INST_IRQ_HAS_IDX(0, 15)
+#define DMA_STM32_0_STREAM_COUNT 16
+#elif DT_INST_IRQ_HAS_IDX(0, 7)
+#define DMA_STM32_0_STREAM_COUNT 8
+#elif DT_INST_IRQ_HAS_IDX(0, 6)
+#define DMA_STM32_0_STREAM_COUNT 7
+#elif DT_INST_IRQ_HAS_IDX(0, 5)
+#define DMA_STM32_0_STREAM_COUNT 6
+#else
+#define DMA_STM32_0_STREAM_COUNT 5
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay) */
```
Can this be cleaned up?

FRASTM replied:
as long as stm32u5 or compatible soc series have one GPDMA of 15 channels

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay)
+#if DT_INST_IRQ_HAS_IDX(1, 7)
+#define DMA_STM32_1_STREAM_COUNT 8
+#elif DT_INST_IRQ_HAS_IDX(1, 6)
+#define DMA_STM32_1_STREAM_COUNT 7
+#elif DT_INST_IRQ_HAS_IDX(1, 5)
+#define DMA_STM32_1_STREAM_COUNT 6
+#else
+#define DMA_STM32_1_STREAM_COUNT 5
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_DRV_INST(1), okay) */
```
Can this be cleaned up?

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+}
+
+/* check if and irq of any type occurred on the channel */
+#define stm32_dma_is_irq_active LL_DMA_IsActiveFlag_MIS
```
Can't we use LL_DMA_IsActiveFlag_MIS directly ?

erwango replied:
This seems unused

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+	return -EAGAIN;
+}
+
+void stm32_dma_set_mem_periph_address(DMA_TypeDef *dma,
+					     uint32_t channel,
+					     uint32_t src_addr,
+					     uint32_t dest_addr)
+{
+	LL_DMA_ConfigAddresses(dma, channel, src_addr, dest_addr);
+}
+
+/* same function to set periph/mem addresses */
+void stm32_dma_set_periph_mem_address(DMA_TypeDef *dma,
+					     uint32_t channel,
+					     uint32_t src_addr,
+					     uint32_t dest_addr)
+{
+	LL_DMA_ConfigAddresses(dma, channel, src_addr, dest_addr);
+}
```
Cant' we use LL api directly ?

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+		}
+		stream->dma_callback(dev, stream->user_data, callback_arg, 0);
+	} else if (stm32_dma_is_tc_irq_active(dma, id)) {
+#ifdef CONFIG_DMAMUX_STM32
```
Is DMAMUX available on U5 ?

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+	return 0;
+}
+
+DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
```
Is DMA_STM32_EXPORT_API required here ?

FRASTM replied:
Since the STM32U5 has _no_ DMAMUX for GPDMA, the DMA_STM32_EXPORT_API is known as 'static'

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 0);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 3);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
```
Can it be simplified?

FRASTM replied:
yes, as long as stm32u5 or compatible soc series have one GPDMA of 15 channels

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+	DMA_STM32_IRQ_CONNECT(0, 2);
+	DMA_STM32_IRQ_CONNECT(0, 3);
+	DMA_STM32_IRQ_CONNECT(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+	DMA_STM32_IRQ_CONNECT(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+	DMA_STM32_IRQ_CONNECT(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+	DMA_STM32_IRQ_CONNECT(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
```
Can it be simplified?

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 3);
+#if DT_INST_IRQ_HAS_IDX(1, 4)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 4);
+#if DT_INST_IRQ_HAS_IDX(1, 5)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 5);
+#if DT_INST_IRQ_HAS_IDX(1, 6)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 6);
+#if DT_INST_IRQ_HAS_IDX(1, 7)
+DMA_STM32_DEFINE_IRQ_HANDLER(1, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(1, 4) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 7) */
```
Can it be simplified?

FRASTM replied:
yes, as long as stm32u5 or compatible soc series have one GPDMA of 15 channels

FRASTM commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,832 @@
+	DMA_STM32_IRQ_CONNECT(1, 1);
+	DMA_STM32_IRQ_CONNECT(1, 2);
+	DMA_STM32_IRQ_CONNECT(1, 3);
+#if DT_INST_IRQ_HAS_IDX(1, 4)
+	DMA_STM32_IRQ_CONNECT(1, 4);
+#if DT_INST_IRQ_HAS_IDX(1, 5)
+	DMA_STM32_IRQ_CONNECT(1, 5);
+#if DT_INST_IRQ_HAS_IDX(1, 6)
+	DMA_STM32_IRQ_CONNECT(1, 6);
+#if DT_INST_IRQ_HAS_IDX(1, 7)
+	DMA_STM32_IRQ_CONNECT(1, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(1, 4) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(1, 7) */
```
yes, as long as stm32u5 or compatible soc series have one GPDMA of 15 channels

FRASTM committed:
boards: arm: stm32u575 nucleo target board has GPDMA instance

Enable the GPDMA 1 instance on the target board nucleo_u575zi_q
from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on 50cf6ca8267e6a87acd3cf1684702245fabc6538
fix merge conflict
change after review

FRASTM commented:
rebase on d01934f9e9d307f9ca97fae15d5dbc8fa2aaceda

FRASTM commented:
check results : twister fails on nucleo_u575zi_q:samples/drivers/watchdog/sample.drivers.watchdog see PR https://github.com/zephyrproject-rtos/zephyr/pull/48062

FRASTM committed:
tests/drivers/dma: testing mem-to-mem DMA transfer on stm32u5

Add the testcase to run on the stm32u5x5 disco kit
and nucleo_u575zi_q boards.
The DMA instance is the GPDMA (up to 16 channels).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on 5897de1e535816247e309348cbe29760361a28b8
fix conflicts
update tests/drivers/dma with a `test_dma` node in each overlay file

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+	DMA_STM32_IRQ_CONNECT(0, 2);
+	DMA_STM32_IRQ_CONNECT(0, 3);
+	DMA_STM32_IRQ_CONNECT(0, 4);
+#if DT_INST_IRQ_HAS_IDX(0, 5)
+	DMA_STM32_IRQ_CONNECT(0, 5);
+#if DT_INST_IRQ_HAS_IDX(0, 6)
+	DMA_STM32_IRQ_CONNECT(0, 6);
+#if DT_INST_IRQ_HAS_IDX(0, 7)
+	DMA_STM32_IRQ_CONNECT(0, 7);
+#endif /* DT_INST_IRQ_HAS_IDX(0, 5) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 6) */
+#endif /* DT_INST_IRQ_HAS_IDX(0, 7) */
```
Are these #idef required today ?

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+/* STM32U5 soc has only one GPDMA instance of 15 channels */
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 0);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 1);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 2);
+DMA_STM32_DEFINE_IRQ_HANDLER(0, 3);
```
Not blocking but use of `LISTIFY` should help reducing this (and make it conditional to a variable number if case happens)

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+		return -EINVAL;
+	}
+
+	dma_stm32_clear_stream_irq(dev, id);
+
+	stm32_dma_enable_stream(dma, id);
```
In this driver what is the difference between these namespaces `dma_stm32` and `stm32_dma`

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+	return -EAGAIN;
+}
+
+void stm32_dma_set_mem_periph_address(DMA_TypeDef *dma,
+					     uint32_t channel,
+					     uint32_t src_addr,
+					     uint32_t dest_addr)
+{
+	LL_DMA_ConfigAddresses(dma, channel, src_addr, dest_addr);
+}
```
I get this is inherited from others drivers, but in case of this specific driver that it mean to be U5 specific, this level abstraction sounds useless

FRASTM replied:
ok for code simplification

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+	return false;
+}
+
+bool stm32_dma_is_unexpected_irq_happened(DMA_TypeDef *dma, uint32_t id)
+{
+	/* Preserve for future amending. */
+	return false;
+}
```
useless

FRASTM replied:
removing

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+			dma_stm32_clear_tc(dma, id);
+		}
+		stream->dma_callback(dev, stream->user_data, callback_arg, 0);
+	} else if (stm32_dma_is_unexpected_irq_happened(dma, id)) {
+		LOG_ERR("Unexpected irq happened.");
+		stream->dma_callback(dev, stream->user_data,
+				     callback_arg, -EIO);
```
Useless code if `stm32_dma_is_unexpected_irq_happened` always returns false

FRASTM replied:
removing

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+	stream->dst_size	= config->dest_data_size;
+
+	/* Check dest or source memory address, warn if 0 */
+	if ((config->head_block->source_address == 0)) {
```
Clean up double parenthesis

FRASTM replied:
done

erwango commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+		LOG_WRN("source_buffer address is null.");
+	}
+
+	if ((config->head_block->dest_address == 0)) {
```
Clean up double parenthesis

FRASTM commented on dts/arm/st/u5/stm32u5.dtsi:
```diff
@@ -437,6 +437,20 @@
+			dma-requests = <114>;
+			dma-offset = <0>;
+			status = "disabled";
+			label = "DMA_1";
```
ok

FRASTM commented on drivers/dma/dma_stm32u5.c:
```diff
@@ -0,0 +1,757 @@
+
+/* STM32U5 soc has only one GPDMA instance of 15 channels */
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
+#define DMA_STM32_0_STREAM_COUNT 16
```
It is through the 
`dma_stm32_streams_##index[DMA_STM32_##index##_STREAM_COUNT`

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -8,13 +8,29 @@ DT_COMPAT_ST_STM32_DMA_V1 := st,stm32-dma-v1
+config DMA_STM32U5
+	bool
+	select USE_STM32_LL_DMA
+	depends on SOC_FAMILY_STM32
```
config DMA_STM32U5 is build like config DMA_STM32 and they are exclusive

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -8,13 +8,29 @@ DT_COMPAT_ST_STM32_DMA_V1 := st,stm32-dma-v1
 DT_COMPAT_ST_STM32_DMAMUX := st,stm32-dmamux
+DT_COMPAT_ST_STM32U5_DMA := st,stm32u5-dma
+
+if DMA
```
DT_COMPAT_ST_STM32U5_DMA  is used to detect which version of DMA driver and choose one of  the config DMA_STM32 or config DMA_STM32U5

erwango replied:
I meant the `if DMA` which should already be done in `drivers/dma/Kconfig`

FRASTM replied:
ok

FRASTM committed:
drivers: dma: stm32 Kconfig with the new DT_HAS_<COMPAT>_ENABLED.

This commit is converting the DMA kconfig for the stm32 dma driver
with new macro DT_HAS_<COMPAT>_ENABLED.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/dma/Kconfig.stm32:
```diff
@@ -8,13 +8,29 @@ DT_COMPAT_ST_STM32_DMA_V1 := st,stm32-dma-v1
 DT_COMPAT_ST_STM32_DMA_V2 := st,stm32-dma-v2
 DT_COMPAT_ST_STM32_DMA_V2BIS := st,stm32-dma-v2bis
 DT_COMPAT_ST_STM32_DMAMUX := st,stm32-dmamux
+DT_COMPAT_ST_STM32U5_DMA := st,stm32u5-dma
```
added to a new commit 898dcbf7697caec7823eb1edb4f699bef2f04675

FRASTM replied:
could also be put in one line 	
`default y if DT_HAS_ST_STM32U5_DMA_ENABLED
`

FRASTM commented:
rebase on e526b407d29d08f61956dcf873f06404b2833018
change the drivers/dma/Kconfig.stm32

FRASTM commented:
rebase on e526b407d29d08f61956dcf873f06404b2833018

erwango commented on drivers/dma/Kconfig.stm32:
```diff
@@ -8,13 +8,25 @@ DT_COMPAT_ST_STM32_DMA_V1 := st,stm32-dma-v1
 	help
-	  DMA driver for STM32 series SoCs.
+	  DMA driver for STM32 series SoCs except STM32U5.
+	  The DMA driver also has a V1 or V2 driver part.
```
"Driver for STM32 DMA V1, V2 and V2bis types".
Let's not mention the not supported types as the list can evolve in the future.

FRASTM replied:
removing that line in the description

erwango replied:
Line is still present. Can you have a check ?

FRASTM replied:
done

mbolivar-nordic reviewed (commented):
DT binding looks good

FRASTM commented:
rebase on f3017aa787e75c77bb7deb9f4ae45d4b4e93fbc3 to fix CI twister error

mbolivar-nordic reviewed (commented):
Gave the DT parts a quick skim and they look OK to me. @teburd I'll leave the +1 on that to you though, since I didn't review in detail

ABOSTM reviewed (approved):
LGTM
